design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+as ARRAY or MAP from columns in KSQL streams or tables into multiple rows. This is the main
+motivation for proposing this feature.
+
+Table funtions would also be useful for other cases such as splitting a string into words, or any
+case where you need to go from a single row to zero or more, for example, unaggregating data or
+duplicating rows.
+
+Here's an example of exploding an array from a column into multiple rows:
+
+We have a stream containing a batch of sensor readings from a particular sensor. For further
+stream processing we want to split them out into separate messages:
+
+Given:
+
+```
+{sensor_id:12345 readings: [23, 56, 3, 76, 75]}
+```
+
+We want to get:
+
+```
+{sensor_id:12345 reading: 23}
+{sensor_id:12345 reading: 56}
+{sensor_id:12345 reading: 3}
+{sensor_id:12345 reading: 76}
+{sensor_id:12345 reading: 75}
+```
+
+This could be accomplished by the following CSAS:
+
+```
+CREATE STREAM exploded_stream AS
+    SELECT sensor_id, EXPLODE(readings) AS reading FROM batched_readings;
+```
+
+Note that any scalar values in the select will be duplicated for each row in the output.
+
+
+## What is in scope
+
+We would like to implement
+
+* Table functions on the SELECT clause only
+* Multiple table functions on the SELECT clause
+* A set of built-in table functions such as EXPLODE (etc)
+* An ability for users to provide their own User Defined Table Functions (UDTFs) using annotations
+in a similar way to how UDFs and UDAFs are currently supported.
+
+## What is not in scope
+
+* Table functions on the FROM clause. We could allow table functions to take the place of other streams
+or tables in the FROM clause. We do not think there is much value in doing this as the main driver
+for this KLIP is in supporting `explode` functionality which is more simply implemented on the
+SELECT clause.
+* Table functions that output more than one column. Some databases support the flattening of an array
+ of N STRUCT types into N rows of M columns, where there is one column for each field of the struct.
+ Instead we propose that flattening an array of N structs will result in N rows with a single column
+ of type STRUCT. The fields of the struct can then be extracted if necessary using a further CSAS
+ which pulls out the struct fields.
+
+## Value/Return
+
+This functionality will make it easier for users to make end-end streaming apps using KSQL
+alone without having to have custom stages using Kafka Streams.
+
+We believe its a common case for users to have array or map like data structures as values
+in messages that they wish to explode. This new feature would allow them to do that in a pure KSQL
+app thus allowing them to create richer KSQL applications.
+
+## Public APIs
+
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
+UDAFs.
```

Comment:
should we have any concept of timing? e.g. if I have an input and it takes me a long time to compute each output, do I need to do it all at once or can I output one row at a time?

Reply:
I could be wrong, but I can't see any way of processing partial results without blocking the caller using Kafka Streams. (See previous comment)

Reply:
As per previous comment ;) I think this may be possible if we use a KS `Processor`.

Reply:
I'm guessing you can't call ProcessorContext.forward for a message after the original call to Processor.process() has returned. Otherwise the upstream processor which is actually reading from the consumer will think the message has been processed before its actually been seen by all downstream processors, and update offsets thus changing delivery semantics (it won't be at least once any more).

Reply:
So you'd need to block processor.process, and that's not going to scale when you've got 1000 queries running in the server. Not to mention the issue you brought up about consumers getting thrown out of consumer groups.